crypto_hal/monocypher/crypto.c

```diff
@@ -0,0 +1,85 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2020 Technology Innovation Institute. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file crypto.c
+ *
+ * Wrapper for the crytpo stuff.
+ *
+ */
+
+#include <inttypes.h>
+#include <stdbool.h>
+
+#include "monocypher/src/optional/monocypher-ed25519.h"
+#include "crypto_hal/monocypher/public_key.h"
+#include "crypto_hal/image_toc.h"
+
+bool verify_app(uint16_t idx, const image_toc_entry_t *toc_entries)
+{
+	volatile uint8_t *app_signature_ptr = NULL;
+	volatile size_t len = 0;
+
+	uint8_t sig_idx = toc_entries[idx].signature_idx;
+	uint8_t sig_key = toc_entries[idx].signature_key;
+
+	const uint8_t *public_key = get_pubkey_by_index(sig_key);
+
+	if (public_key == NULL) {
+		return false;
+	}
+
+	app_signature_ptr = (volatile uint8_t *)toc_entries[sig_idx].start;
+	len = (size_t)toc_entries[idx].end - (size_t)toc_entries[idx].start;
```

Comment:
Should we check that this does not exceed the flash size?
What happens if len is 0?

Reply:
I was thinking that runtime checks are not needed any more here;
The TOC is a constant table, and the contents have been verified to be as the originator/vendor intended since the TOC is withing the first app, and the signature for that is always checked.

I doesn't harm to have have the checks here, except for a few words of flash being used. I can add the checks if you feel like it, or I can also just comment this?

Reply:
> The TOC is a constant table, and the contents have been verified to be as the originator/vendor intended since the TOC is within the first app, and the signature for that is always checked.

As `verify_app` is also used to validate the first entry, the TOC cannot be trusted here yet. Or do I miss something?

Reply:
I had put the check for the first app start/end boundaries into the code that verifies that the TOC itself is valid in bool find_toc()

Thanks for looking it through, please let me know if you think that there is something wrong with this approach!

Reply:
And the reason for not doing it for every entry in the TOC is, that  I didn't want to limit the entries to be within the internal flash; there could be other memories, e.g. one TOC entry can initiate a copy/decrypt to SRAM and the later one check it from there....

Reply:
Yes I've seen those but they are not enough. Specifically, the first entry must be checked for:
- sig_idx bound check
- `len` could be 0 (if you confirm this fails the crypto_ed25519_check check, it's enough to add a comment) or even wrap around as it's unsigned and it's possible to pass `end` < `start` (which passes the check in `find_toc`). An additional check for `end > start` fixes both.

Reply:
These are valid points, even though I don't see how it could pass the signature check if the sig_idx or len fields would be forged, this is the type of thing which should be fixed. Especially if someone uses the common TOC code with some other crypto algorithms.

Thanks again, I will add end>start and sig_idx bounds checks!

Reply:
I updated the find_toc function;

While checking the sig_idx, I thing the following should also be taken into account:
- The values from board_info structure are not trustworthy, they come along the app
- If checking the sig_idx, also the start and end addresses of the signature should  be checked the same way as for the app itself.
Hopefully I got it right this time...